File: c/src/args.c
#include "args.h"

#include <errno.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

#define DEFAULT_PORTS "1-1024"
#define DEFAULT_TIMEOUT_MS 800u
#define DEFAULT_CONCURRENCY 500u
#define MIN_TIMEOUT_MS 50u
#define MAX_TIMEOUT_MS 60000u
#define MIN_CONCURRENCY 1u
#define MAX_CONCURRENCY 10000u

static void set_err(char *err, size_t err_len, const char *text) {
    if (err == NULL || err_len == 0) {
        return;
    }
    snprintf(err, err_len, "%s", text);
}

static int parse_ulong(const char *text, unsigned long *out) {
    char *end = NULL;

    if (text == NULL || *text < '0' || *text > '9') {
        return 0;
    }
    errno = 0;
    *out = strtoul(text, &end, 10);
    return errno == 0 && *end == '\0';
}

static int is_value_flag(const char *arg) {
    return strcmp(arg, "--ports") == 0 || strcmp(arg, "--timeout") == 0 ||
           strcmp(arg, "--concurrency") == 0 || strcmp(arg, "--format") == 0;
}

static ArgsStatus apply_number(const char *flag, const char *value, Options *out, char *err, size_t err_len) {
    unsigned long number = 0;

    if (!parse_ulong(value, &number)) {
        set_err(err, err_len, "a number was expected");
        return ARGS_ERROR;
    }
    if (strcmp(flag, "--timeout") == 0) {
        if (number < MIN_TIMEOUT_MS || number > MAX_TIMEOUT_MS) {
            set_err(err, err_len, "timeout must be between 50 and 60000 ms");
            return ARGS_ERROR;
        }
        out->timeout_ms = (unsigned)number;
    } else {
        if (number < MIN_CONCURRENCY || number > MAX_CONCURRENCY) {
            set_err(err, err_len, "concurrency must be between 1 and 10000");
            return ARGS_ERROR;
        }
        out->concurrency = (unsigned)number;
    }
    return ARGS_OK;
}

static ArgsStatus apply_value(const char *flag, const char *value, Options *out, char *err, size_t err_len) {
    if (strcmp(flag, "--ports") == 0) {
        if (strlen(value) >= sizeof(out->ports_spec)) {
            set_err(err, err_len, "port list is too long");
            return ARGS_ERROR;
        }
        snprintf(out->ports_spec, sizeof(out->ports_spec), "%s", value);
        return ARGS_OK;
    }
    if (strcmp(flag, "--format") == 0) {
        if (strcmp(value, "table") == 0) {
            out->format = FORMAT_TABLE;
        } else if (strcmp(value, "json") == 0) {
            out->format = FORMAT_JSON;
        } else {
            set_err(err, err_len, "format must be table or json");
            return ARGS_ERROR;
        }
        return ARGS_OK;
    }
    return apply_number(flag, value, out, err, err_len);
}

void args_defaults(Options *out) {
    memset(out, 0, sizeof(*out));
    snprintf(out->ports_spec, sizeof(out->ports_spec), "%s", DEFAULT_PORTS);
    out->timeout_ms = DEFAULT_TIMEOUT_MS;
    out->concurrency = DEFAULT_CONCURRENCY;
    out->format = FORMAT_TABLE;
}

const char *args_usage(void) {
    return "portscan <target> [--ports 1-1024|22,80,443] [--timeout <ms>] [--concurrency <n>]\n"
           "                 [--banner] [--format table|json] [--yes-i-own-this] [--help]\n";
}

static ArgsStatus take_positional(const char *arg, Options *out, char *err, size_t err_len) {
    if (out->target[0] != '\0') {
        set_err(err, err_len, "only one target is allowed");
        return ARGS_ERROR;
    }
    if (strlen(arg) >= sizeof(out->target)) {
        set_err(err, err_len, "target is too long");
        return ARGS_ERROR;
    }
    snprintf(out->target, sizeof(out->target), "%s", arg);
    return ARGS_OK;
}

ArgsStatus args_parse(int argc, char **argv, Options *out, char *err, size_t err_len) {
    int i;

    args_defaults(out);

    for (i = 1; i < argc; i++) {
        const char *arg = argv[i];
        ArgsStatus status = ARGS_OK;

        if (strcmp(arg, "--help") == 0 || strcmp(arg, "-h") == 0) {
            return ARGS_HELP;
        }
        if (strcmp(arg, "--banner") == 0) {
            out->banner = 1;
        } else if (strcmp(arg, "--yes-i-own-this") == 0) {
            out->confirmed = 1;
        } else if (is_value_flag(arg)) {
            if (i + 1 >= argc) {
                set_err(err, err_len, "a value is missing");
                return ARGS_ERROR;
            }
            status = apply_value(arg, argv[++i], out, err, err_len);
        } else if (strncmp(arg, "--", 2) == 0) {
            set_err(err, err_len, "unknown argument");
            return ARGS_ERROR;
        } else {
            status = take_positional(arg, out, err, err_len);
        }
        if (status != ARGS_OK) {
            return status;
        }
    }

    if (out->target[0] == '\0') {
        set_err(err, err_len, "a target is required");
        return ARGS_ERROR;
    }
    return ARGS_OK;
}

File: csharp/src/PortScan/Args.cs
namespace PortScan;

public enum ReportFormat
{
    Table,
    Json,
}

public sealed record Options
{
    public const int DefaultTimeoutMs = 800;
    public const int MinTimeoutMs = 50;
    public const int MaxTimeoutMs = 60_000;
    public const int DefaultConcurrency = 500;
    public const int MinConcurrency = 1;
    public const int MaxConcurrency = 10_000;

    public string Target { get; init; } = string.Empty;

    public string PortsSpec { get; init; } = "1-1024";

    public TimeSpan Timeout { get; init; } = TimeSpan.FromMilliseconds(DefaultTimeoutMs);

    public int Concurrency { get; init; } = DefaultConcurrency;

    public bool Banner { get; init; }

    public ReportFormat Format { get; init; } = ReportFormat.Table;

    public bool Confirmed { get; init; }

    public bool Help { get; init; }
}

public sealed class ArgumentParseException : Exception
{
    public ArgumentParseException(string message)
        : base(message)
    {
    }

    public ArgumentParseException()
    {
    }

    public ArgumentParseException(string message, Exception innerException)
        : base(message, innerException)
    {
    }
}

public static class Args
{
    public static string Usage =>
        "portscan <target> [--ports 1-1024|22,80,443] [--timeout <ms>] [--concurrency <n>]" + Environment.NewLine +
        "                 [--banner] [--format table|json] [--yes-i-own-this] [--help]" + Environment.NewLine;

    private static readonly HashSet<string> ValueFlags = ["--ports", "--timeout", "--concurrency", "--format"];

    public static Options Parse(IReadOnlyList<string> arguments)
    {
        ArgumentNullException.ThrowIfNull(arguments);

        var options = new Options();

        for (var index = 0; index < arguments.Count; index++)
        {
            var argument = arguments[index];

            if (argument is "--help" or "-h")
            {
                return options with { Help = true };
            }
            if (argument == "--banner")
            {
                options = options with { Banner = true };
                continue;
            }
            if (argument == "--yes-i-own-this")
            {
                options = options with { Confirmed = true };
                continue;
            }
            if (ValueFlags.Contains(argument))
            {
                if (index + 1 >= arguments.Count)
                {
                    throw new ArgumentParseException($"{argument} is missing a value");
                }
                options = ApplyValue(argument, arguments[++index], options);
                continue;
            }
            if (argument.StartsWith("--", StringComparison.Ordinal))
            {
                throw new ArgumentParseException($"unknown argument: {argument}");
            }
            if (options.Target.Length != 0)
            {
                throw new ArgumentParseException("only one target is allowed");
            }
            options = options with { Target = argument };
        }

        if (options.Target.Length == 0)
        {
            throw new ArgumentParseException("a target is required");
        }
        return options;
    }

    private static Options ApplyValue(string flag, string value, Options options) => flag switch
    {
        "--ports" => options with { PortsSpec = value },
        "--format" => options with { Format = ParseFormat(value) },
        "--timeout" => options with { Timeout = TimeSpan.FromMilliseconds(ParseBoundedNumber(value, Options.MinTimeoutMs, Options.MaxTimeoutMs, "timeout must be between 50 and 60000 ms")) },
        _ => options with { Concurrency = ParseBoundedNumber(value, Options.MinConcurrency, Options.MaxConcurrency, "concurrency must be between 1 and 10000") },
    };

    private static ReportFormat ParseFormat(string value) => value switch
    {
        "table" => ReportFormat.Table,
        "json" => ReportFormat.Json,
        _ => throw new ArgumentParseException("format must be table or json"),
    };

    private static int ParseBoundedNumber(string value, int minimum, int maximum, string errorMessage)
    {
        if (!long.TryParse(value, System.Globalization.NumberStyles.None, System.Globalization.CultureInfo.InvariantCulture, out var number) ||
            number < minimum || number > maximum)
        {
            throw new ArgumentParseException(errorMessage);
        }
        return (int)number;
    }
}
